raft topology: support --ignore-dead-nodes in removenode and replace

patjed41 opened the pull request:
We add support for `--ignore-dead-nodes` in `raft_removenode` and `--ignore-dead-nodes-for-replace` in `raft_replace`. For now, we allow passing only host ids of the ignored nodes. Supporting IPs is currently impossible because `raft_address_map` doesn't provide a mapping from IP to a host id.

The main steps of the implementation are as follows:
- add the `ignore_nodes` column to `system.topology`,
- set the `ignore_nodes` value of the topology mutation in `raft_removenode` and `raft_replace`,
- extend `service::request_param` with alternative types that allow storing a set of ids of the ignored nodes,
- load `ignore_nodes` from `system.topology` into `request_param` in `system_keyspace::load_topology_state`,
- add `ignore_nodes` to `exclude_nodes` in `topology_coordinator::exec_global_command`,
- pass `ignore_nodes` to `replace_with_repair` and `remove_with_repair` in `storage_service::raft_topology_cmd_handler`.

Additionally, we add `test_raft_ignore_nodes.py` with two tests that verify the added changes.

Fixes #15025

kbr-scylla commented on db/system_keyspace.cc:
```diff
@@ -2526,6 +2535,11 @@ future<service::topology> system_keyspace::load_topology_state() {
             rebuild_option = row.get_as<sstring>("rebuild_option");
         }
 
+        std::unordered_set<raft::server_id> ignored_ids;
```
> In system_keyspace::load_topology_state, we don't need to emplace
ignored_ids into ret.req_param if a row contains topology_request.
This parameter will only be used after deleting topology_request.

But I see that you emplace it unconditionally nevertheless. Is that paragraph of the commit message important in any way?

gleb-cloudius replied:
The code emplaces an empty set in this case, but why the complication?

patjed41 replied:
> The code emplaces an empty set in this case, but why the complication?

Right, it's not worth doing it. I changed it to emplace `ignored_ids` for the replace operation in both cased.

The paragraph still applies to the removenode operation, but I think it brings more confusion than help, so I removed it.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -960,6 +960,16 @@ class topology_coordinator {
+        if (node.rs->state == node_state::replacing) {
+            return std::get<replace_param>(node.req_param.value()).ignored_ids;
+        }
+        return {};
```
Maybe we should do `on_internal_error` here for now since we're not supposed to call `parse_ignore_nodes` in situations other than `removing`/`replacing`?

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -5747,12 +5747,20 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
-                                co_await _repair.local().replace_with_repair(get_token_metadata_ptr(), rs.ring.value().tokens, {});
+                                std::unordered_set<gms::inet_address> ignored_ips;
+                                for (const auto& id : std::get<replace_param>(_topology_state_machine._topology.req_param[raft_server.id()]).ignored_ids) {
+                                    auto ip = _group0->address_map().find(id);
```
Please live a FIXME with a reference to #6403 (we should not need to do this translation here)

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -5780,20 +5788,29 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
+                            assert(_topology_state_machine._topology.req_param.contains(id));
+                            std::list<gms::inet_address> ignored_ips;
+                            for (const auto& ignored_id : std::get<std::unordered_set<raft::server_id>>(_topology_state_machine._topology.req_param[id])) {
+                                auto ip = _group0->address_map().find(ignored_id);
```
Same FIXME as above

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -5780,20 +5788,29 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
                         if (is_repair_based_node_ops_enabled(streaming::stream_reason::removenode)) {
-                            auto ops = seastar::make_shared<node_ops_info>(node_ops_id::create_random_id(), as, std::list<gms::inet_address>());
-                            // FIXME: ignore node list support
+                            assert(_topology_state_machine._topology.req_param.contains(id));
```
Use `on_internal_error` instead

patjed41 replied:
Done.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -5747,12 +5747,20 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
                         // Bootstrap did not complete yet, but streaming did
                     } else {
                         co_await retrier(_bootstrap_result, coroutine::lambda([&] () ->future<> {
+                            assert(_topology_state_machine._topology.req_param.contains(raft_server.id()));
```
Use `on_internal_error` instead

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3174/

patjed41 commented on service/storage_service.cc:
```diff
@@ -2079,6 +2084,19 @@ future<> storage_service::raft_state_monitor_fiber(raft::server& raft, cdc::gene
+    std::unordered_set<raft::server_id> ignored_ids;
+    for (const sstring& n : ignore_nodes_strs) {
+        raft::server_id ignored_id;
+        // FIXME: IPs should be supported, but the current implementation of raft_address_map doesn't allow this.
```
Sure, but I'd like to know if there is any expected solution for this FIXMEs to provide more information in the issue ticket. Is it simply adding a new member function -- `find_by_inet_address` -- to `raft_address_map` together with some unit tests?

By the way, I left two TODOs in `test_raft_ignore_nodes.py` that depend on these FIXMEs and should be done together with them.

kbr-scylla replied:
> Is it simply adding a new member function -- find_by_inet_address -- to raft_address_map together with some unit tests?

That is one of the alternatives.
Although we should probably do the translation the same way regular removenode does, IIRC it's done using `locator::topology` somehow.

Another alternative is to update the documentation that we no longer support passing IP addresses to this option. However that's a bit problematic because there was no deprecation period.

Best would be a mix of both: update the documentation that passing IPs is deprecated, and do the translation here. Two releases later we would update the documentation that passing IPs is no longer supported and remove the translation.

patjed41 replied:
I created the follow-up issue -- #15126.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -2079,6 +2084,19 @@ future<> storage_service::raft_state_monitor_fiber(raft::server& raft, cdc::gene
+    std::unordered_set<raft::server_id> ignored_ids;
+    for (const sstring& n : ignore_nodes_strs) {
+        raft::server_id ignored_id;
+        // FIXME: IPs should be supported, but the current implementation of raft_address_map doesn't allow this.
```
We can also deprecate the use of IPs with consistent topology.

patjed41 replied:
I added this to the FIXME description.

gleb-cloudius reviewed (approved):
LGTM. But missing the isolation of ignored nodes. Not a merge blocker, but a TODO item that needs to be addressed at some point.

patjed41 commented on service/storage_service.cc:
```diff
@@ -2089,14 +2107,20 @@ future<> storage_service::raft_replace(raft::server& raft_server, raft::server_i
 
+        for (const auto& ignored_id : ignored_ids) {
+            if (!_topology_state_machine._topology.find(ignored_id)) {
+                throw std::runtime_error(::format("raft topology replace: ignored node {} is not found in the cluster", ignored_id));
```
> Is this the same behavior as for old replace?

I think so. The old replace also throws `runtime_error` if an ignored node is not in the cluster (see `storage_service::parse_node_list`). Only the error message is different.

> BTW let's remove "raft topology " prefix from errors that we throw through user APIs ("raft topology" is an implementation detail). It's fine to keep it in log messages.

I changed the errors added by me. What about the ones already present in `raft_removenode`? Should I fix them in another PR, since they are unrelated?

kbr-scylla replied:
> Should I fix them in another PR, since they are unrelated?

Sure, after this PR is merged.

patjed41 commented on service/topology_state_machine.hh:
```diff
@@ -46,7 +46,12 @@ enum class topology_request: uint16_t {
 };
 
-using request_param = std::variant<raft::server_id, sstring, uint32_t>;
+struct replace_param {
```
> Can you also create a struct for the `std::unordered_set<raft::server_id>`?

Done.

> As a follow-up, we should also create structs for the `sstring` and `uint32_t`.

I can work on that.

patjed41 commented:
v2:
- introduced `service::removenode_param`
- removed "raft topology" prefixes from the errors thrown in `raft_removenode` and `raft_replace`
- changed `assert`s to `on_internal_error`s in `storage_service::raft_topology_cmd_handler`
- added 2 FIXMEs in `storage_service::raft_topology_cmd_handler` about `id -> IP` translation
- changed descriptions of the FIXMEs in `raft_removenode` and `raft_replace`
- fixed some typos

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3194/
